File: python/example/sort/service/quicksort_service.py
from otterop.lang.array import Array
from otterop.lang.string import String

class QuicksortService:

    def __init__(self):
        pass
    def swap(self, array, from_idx, to_idx):
        tmp = array.get(to_idx)
        array.set(to_idx, array.get(from_idx))
        array.set(from_idx, tmp)

    def sort_with_indices(self, array, from_idx, to_idx):
        i = from_idx
        swap_with = -1
        pivot = to_idx - 1
        while i < pivot:
            i_string = array.get(i)
            pivot_string = array.get(pivot)
            less_than = i_string.compare_to(pivot_string) < 0
            if swap_with >= 0 and less_than:
                self.swap(array, swap_with, i)
                swap_with = swap_with + 1
            elif swap_with < 0 and not less_than:
                swap_with = i
            i = i + 1
        if swap_with >= 0 and swap_with < pivot:
            swap_with_string = array.get(swap_with)
            pivot_string = array.get(pivot)
            if swap_with_string.compare_to(pivot_string) >= 0:
                self.swap(array, swap_with, pivot)
                pivot = swap_with
        if pivot > from_idx + 1:
            self.sort_with_indices(array, from_idx, pivot)
        if pivot + 1 < to_idx - 1:
            self.sort_with_indices(array, pivot + 1, to_idx)

    def sort(self, array):
        self.sort_with_indices(array, 0, array.size())
        return array

File: go/example/sort/service/quicksortservice.go
package service

import (
    lang "github.com/otterop/otterop/go/lang"
)


type QuicksortService struct {
}




func QuicksortServiceNew() *QuicksortService {
    this := new(QuicksortService)
    return this
}

func (this *QuicksortService) swap(array *lang.Array[*lang.String], fromIdx int, toIdx int)  {
    var tmp *lang.String = array.Get(toIdx)
    array.Set(toIdx, array.Get(fromIdx))
    array.Set(fromIdx, tmp)
}

func (this *QuicksortService) sortWithIndices(array *lang.Array[*lang.String], fromIdx int, toIdx int)  {
    var i int = fromIdx
    var swapWith int = -1
    var pivot int = toIdx - 1
    
    for i < pivot {
        var iString *lang.String = array.Get(i)
        var pivotString *lang.String = array.Get(pivot)
        var lessThan bool = iString.CompareTo(pivotString) < 0
        
        if swapWith >= 0 && lessThan {
            this.swap(array, swapWith, i)
            swapWith = swapWith + 1
        } else if swapWith < 0 && !lessThan {
            swapWith = i
        }
        i = i + 1
    }
    
    if swapWith >= 0 && swapWith < pivot {
        var swapWithString *lang.String = array.Get(swapWith)
        var pivotString *lang.String = array.Get(pivot)
        
        if swapWithString.CompareTo(pivotString) >= 0 {
            this.swap(array, swapWith, pivot)
            pivot = swapWith
        }
    }
    
    if pivot > fromIdx + 1 {
        this.sortWithIndices(array, fromIdx, pivot)
    }
    
    if pivot + 1 < toIdx - 1 {
        this.sortWithIndices(array, pivot + 1, toIdx)
    }
}

func (this *QuicksortService) Sort(array *lang.Array[*lang.String]) *lang.Array[*lang.String] {
    this.sortWithIndices(array, 0, array.Size())
    return array
}